### PR TITLE
Backfill all accounts during beam sync

### LINKED ISSUE #1846: PeerConnectionLost during stream_transport_messages
### What is wrong?

Similar to but distinct from #1767 :
```
   DEBUG  07-06 22:15:20  async_process_runner.py  b'  Traceback (most recent call last):\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/transport.py", line 236, in read\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      return await asyncio.wait_for(self._reader.readexactly(n), timeout=CONN_IDLE_TIMEOUT)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/usr/local/lib/python3.7/asyncio/tasks.py", line 442, in wait_for\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      return fut.result()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/usr/local/lib/python3.7/asyncio/streams.py", line 677, in readexactly\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      raise IncompleteReadError(incomplete, n)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'  asyncio.streams.IncompleteReadError: 0 bytes read on a total of 32 expected bytes\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'  The above exception was the direct cause of the following exception:\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'  Traceback (most recent call last):\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/.tox/py37-wheel-cli/lib/python3.7/site-packages/async_service/base.py", line 300, in _run_and_manage_task\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      await task.run()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/.tox/py37-wheel-cli/lib/python3.7/site-packages/async_service/asyncio.py", line 35, in run\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      await self._async_fn(*self._async_fn_args)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/connection.py", line 187, in _feed_protocol_handlers\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      async for cmd in self._multiplexer.stream_protocol_messages(protocol):\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 273, in _stream_protocol_messages\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      self.raise_if_streaming_error()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 314, in raise_if_streaming_error\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      raise err\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/.tox/py37-wheel-cli/lib/python3.7/site-packages/async_service/base.py", line 300, in _run_and_manage_task\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      await task.run()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/.tox/py37-wheel-cli/lib/python3.7/site-packages/async_service/asyncio.py", line 35, in run\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      await self._async_fn(*self._async_fn_args)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/connection.py", line 187, in _feed_protocol_handlers\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      async for cmd in self._multiplexer.stream_protocol_messages(protocol):\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 273, in _stream_protocol_messages\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      self.raise_if_streaming_error()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 314, in raise_if_streaming_error\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      raise err\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 328, in _do_multiplexing\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      await self._handle_commands(msg_stream)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 340, in _handle_commands\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      async for protocol, cmd in msg_stream:\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/multiplexer.py", line 56, in stream_transport_messages\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      msg = await transport.recv()\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/transport.py", line 244, in recv\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      header_bytes = await self.read(HEADER_LEN + MAC_LEN)\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'    File "/home/circleci/repo/p2p/transport.py", line 238, in read\n'
   DEBUG  07-06 22:15:20  async_process_runner.py  b'      raise PeerConnectionLost(f"Lost connection to {self.remote}") from err\n'
```
from [this PR run](https://app.circleci.com/pipelines/github/ethereum/trinity/6664/workflows/ce41e820-ff65-4adc-937f-f74836b0fa3f/jobs/250463/steps).

#1750 seems to touch similar code, so maybe it's related?

### How can it be fixed

:man_shrugging: I guess catch the connection lost in `Multiplexer._do_multiplexing()` next to the `TimeoutError`. It doesn't seem like a big problem, and that we can just stop streaming the messages, but I may be missing something:

https://github.com/ethereum/trinity/blob/e7f3a277c0fd0c8fbaf16e5ed078cd9e68273afd/p2p/multiplexer.py#L327-L334

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

Related to #854 

### How was it fixed?

- Includes test of account backfill with cold state database
- Does not include bytecode or storage backfill

(Skipped bytecode/storage to make the PR a bit more manageable)

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)
- [x] Fixes #1846 (add release note)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by carver** [comment]
Last three commits are totally unrelated, just handling some CI issues that popped up.

File: trinity/sync/beam/backfill.py [hunk 1]
```diff
@@ -77,44 +100,184 @@ async def run(self) -> None:
...
+        if self._account_tracker.is_complete:
+            return True
+        else:
+            # At least one account trie node is missing
+            return False
 
-            self.manager.run_task(self._make_request, peer, on_deck)
+    async def _get_next_node(self) -> AsyncIterator[TrackedRequest]:
+        """
+        yields one account node which needs to be downloaded
```
Comment (cburgdorf) [inline trinity/sync/beam/backfill.py:140]: Maybe:

>Walk the whole account trie and yield nodes to be downloaded until no more nodes are missing.

File: trinity/sync/beam/backfill.py [hunk 2]
```diff
@@ -77,44 +100,184 @@ async def run(self) -> None:
...
+            except trie_exceptions.TraversedPartialPath as exc:
+                node = exc.simulated_node
+
+            if node.value:
+                full_key_nibbles = path_to_node + node.suffix
+
+                if len(node.sub_segments):
+                    # It shouldn't be a problem to skip handling this case, because all keys are
+                    #   hashed 32 bytes.
+                    raise Exception(
```
Comment (cburgdorf) [inline trinity/sync/beam/backfill.py:241]: The bare `Exception` is a bit unexpected. Is this intentional?

File: trinity/sync/beam/backfill.py [hunk 3]
```diff
@@ -129,85 +292,22 @@ def _insert_results(
...
-            # leaf or extension node
-            return {decoded_node[1]}
-        else:
-            # final value, ignore
-            return set()
-
-    def set_root_hash(self, root_hash: Hash32) -> None:
-        if len(self._node_hashes) < REQUEST_SIZE:
-            self._node_hashes.append(root_hash)
+    def set_root_hash(self, header: BlockHeader, root_hash: Hash32) -> None:
```
Comment (cburgdorf) [inline trinity/sync/beam/backfill.py:298]: `BlockHeader` -> `BlockHeaderAPI`

File: trinity/sync/beam/backfill.py [hunk 4]
```diff
@@ -129,85 +292,22 @@ def _insert_results(
...
+        if self._next_trie_root_hash is None:
+            self._next_trie_root_hash = root_hash
+            self._begin_backfill.set()
+        elif header.block_number % 32 == 1:
```
Comment (cburgdorf) [inline trinity/sync/beam/backfill.py:302]: I read this as the state backfill pivots to a new `state_root` every `32` blocks. If that's the case then I think the `32` should also go into `constants.py`

File: trinity/sync/beam/backfill.py [hunk 5]
```diff
@@ -77,44 +100,184 @@ async def run(self) -> None:
...
-                continue
+    def _is_complete(self) -> bool:
+        if self._account_tracker.is_complete:
+            return True
+        else:
+            # At least one account trie node is missing
+            return False
 
-            self.manager.run_task(self._make_request, peer, on_deck)
+    async def _get_next_node(self) -> AsyncIterator[TrackedRequest]:
```
[the remark this answers is not in the document]
Comment (carver) [reply trinity/sync/beam/backfill.py:138]: Haha, yeah, you've caught that once or twice before. Hopefully you'll only have to catch that once or twice more, again. :laughing: Apparently, it's a pretty ingrained bad habit I have for naming.

**Pull request merged**